File: Python/day14/main.py
ROUND = "O".encode()[0]
SQUARE = "#".encode()[0]
EMPTY = ".".encode()[0]


def load_data(filename: str) -> list[bytearray]:
    with open(filename, encoding="UTF-8") as f:
        grid = []
        for line in f.readlines():
            grid.append(bytearray(line.strip().encode()))
        return grid


def tilt_north(grid: list[bytearray]):
    for row in range(len(grid)):
        for col in range(len(grid)):
            if grid[row][col] == ROUND:
                current_row = row
                while current_row > 0 and grid[current_row - 1][col] == EMPTY:
                    current_row -= 1
                if current_row != row:
                    grid[row][col] = EMPTY
                    grid[current_row][col] = ROUND


def calculate_load(grid: list[bytearray]) -> int:
    load = 0
    dim = len(grid)
    for row in range(dim):
        for col in range(dim):
            if grid[row][col] == ROUND:
                load += (dim - row)
    return load


def rotate_grid(grid: list[bytearray]):
    dim = len(grid)
    for row in range(dim):
        for col in range(row, dim):
            grid[row][col], grid[col][row] = grid[col][row], grid[row][col]
    for col in range(dim // 2):
        for row in range(dim):
            grid[row][col], grid[row][dim - col - 1] = grid[row][dim - col - 1], grid[row][col]


def print_grid(grid: list[bytearray]):
    for row in grid:
        print(row)
    print()


def grid_to_string(grid: list[bytearray]):
    return "".join([row.decode() for row in grid])


def solution_part_1():
    grid = load_data("input_example.txt")
    tilt_north(grid)
    print(f"Load: {calculate_load(grid)}")


def cycle(grid: list[bytearray]):
    for _ in range(4):
        tilt_north(grid)
        rotate_grid(grid)


def find_cycle(grid: list[bytearray]) -> tuple[int, int]:
    cache = {grid_to_string(grid): [0]}
    index = 0
    while True:
        cycle(grid)
        indices = cache.get(grid_to_string(grid), [])
        indices.append(index)
        cache[grid_to_string(grid)] = indices
        if len(indices) == 2:
            return indices[1] - indices[0], (index + 1)
        index += 1


def solution_part_2():
    grid = load_data("input.txt")
    remaining_cycles = 1_000_000_000
    cycle_len, current_cycles = find_cycle(grid)
    remaining_cycles -= current_cycles
    remaining_cycles -= (remaining_cycles // cycle_len) * cycle_len
    for _ in range(remaining_cycles):
        cycle(grid)
    print(f"Load: {calculate_load(grid)}")


def main():
    solution_part_1()
    solution_part_2()


if __name__ == "__main__":
    main()

File: Rust/day14/src/main.rs
use std::collections::{HashMap};

fn load_data(filename: &str) -> Vec<u8> {
    let data = std::fs::read_to_string(filename).unwrap();
    let mut platform = Vec::new();
    for line in data.lines() {
        let mut row = Vec::new();
        for c in line.chars() {
            match c {
                '.' => row.push(0),
                '#' => row.push(1),
                'O' => row.push(2),
                _ => panic!("Unknown field type")
            }
        }
        platform.push(row);
    }
    platform.iter().flat_map(|row| row.iter()).copied().collect()
}

fn roll_north(platform: &mut [u8], dimension: usize) {
    for y in 0..dimension {
        for x in 0..dimension {
            if platform[y * dimension + x] == 2 {
                let mut current_row = y;
                while current_row > 0 && platform[(current_row - 1) * dimension + x] == 0 {
                    current_row -= 1;
                }
                if current_row != y {
                    platform[current_row * dimension + x] = 2;
                    platform[y* dimension + x] = 0;
                }
            }
        }
    }
}

fn calculate_load(platform: &[u8], dimension: usize) -> usize {
    let mut load = 0;
    for y in 0..dimension {
        for x in 0..dimension {
            if platform[y * dimension + x] == 2 {
                load += dimension - y;
            }
        }
    }
    load
}

#[allow(dead_code)]
fn print_field(platform: &[u8], dimension: usize) {
    for y in 0..dimension {
        for x in 0..dimension {
            print!("{:?}", platform[y * dimension + x]);
        }
        println!();
    }
}

fn rotate_field(platform: &mut [u8], dimension: usize) {
    for i in 0..dimension {
        for j in i..dimension {
            platform.swap(i * dimension + j, j * dimension + i);
        }
    }
    for i in 0..dimension {
        for j in 0..dimension / 2 {
            platform.swap(i * dimension + j, i * dimension + dimension - j - 1);
        }
    }
}

fn solution_part_1() {
    let mut field = load_data("input.txt");
    let dimension = (field.len() as f64).sqrt() as usize;
    roll_north(&mut field, dimension);
    let load = calculate_load(&field, dimension);
    println!("Load: {}", load);
}

fn cycle_field(field: &mut Vec<u8>, dimension: usize) {
    for _ in 0..4 {
        roll_north(field, dimension);
        rotate_field(field, dimension);
    }
}

fn find_cycle(field: &mut Vec<u8>, dimension: usize) -> (usize, usize) {
    let mut cache: HashMap<Vec<u8>, Vec<usize>> = HashMap::new();
    cache.insert(field.clone(), vec![0]);
    for i in 0.. {
        cycle_field(field, dimension);
        let indices = cache.entry(field.clone()).or_default();
        indices.push(i + 1);
        if indices.len() == 2 {
            return (indices[1] - indices[0], i + 1);
        }
    }
    unreachable!();
}

fn solution_part_2() {
    let mut field = load_data("input.txt");
    let dimension = (field.len() as f64).sqrt() as usize;
    let mut remaining_cycles = 1000000000;
    let (cycle, current_cycles) = find_cycle(&mut field, dimension);
    remaining_cycles -= current_cycles;
    let cycles = remaining_cycles / cycle;
    remaining_cycles -= cycles * cycle;
    for _ in 0..remaining_cycles {
        cycle_field(&mut field, dimension);
    }
    let load = calculate_load(&field, dimension);
    println!("Load: {}", load);
}
fn main() {
    solution_part_1();
    solution_part_2();
}
